builder/builder.go

```diff
@@ -155,6 +160,52 @@ func (b *Builder) BuildTarget(ctx context.Context, target domain.Target, opt Bui
 	return mts, nil
 }
 
+func (b *Builder) startRegistryProxy(ctx context.Context, caps apicaps.CapSet) (func(), bool) {
+
+	if !b.opt.UseRemoteRegistry {
+		b.opt.Console.VerbosePrintf("Registry proxy not enabled")
+		return nil, false
+	}
+
+	if err := caps.Supports(pb.CapEarthlyRegistryProxy); err != nil {
+		b.opt.Console.Warnf("Registry proxy is not supported by BuildKit: %s", err)
+		return nil, false
+	}
+
+	addr := "localhost:0" // Have the OS select a port
+
+	lnCfg := net.ListenConfig{}
+	ln, err := lnCfg.Listen(ctx, "tcp", addr)
+	if err != nil {
+		b.opt.Console.Warnf("Failed to create proxy listener: %v", err)
+		return nil, false
+	}
+
+	p := regproxy.NewRegistryProxy(ln, b.s.bkClient.RegistryClient())
+	go p.Serve(ctx)
+
+	addr = fmt.Sprintf("localhost:%d", ln.Addr().(*net.TCPAddr).Port)
+	b.opt.Console.VerbosePrintf("Starting local registry proxy: %s", addr)
+
+	doneCh := make(chan struct{})
+
+	go func() {
+		for err := range p.Err() {
+			if err != nil && !errors.Is(err, context.Canceled) {
+				b.opt.Console.Warnf("Failed to serve registry proxy: %v", err)
+			}
+		}
+		doneCh <- struct{}{}
+	}()
+
+	b.opt.LocalRegistryAddr = addr
+
+	return func() {
+		p.Close()
+		<-doneCh
```

Comment:
Should this listen to `ctx.Done()` too while waiting for `<-doneCh`?

Reply:
That's probably worthwhile. `p.Close` will kill the server, but I don't think it's immediate.